### ISSUE TITLE
Loss function and optimization 

### ISSUE THREAD

**Issue by EDW6** [EDW6/StudyCv#2]
Recall for last time: challenges of recognition, viewpoint, illumination, deformation, occlusion, clutter, intraclass variation, 
Data-driven approach, KNN； linear classifier, f(x,w), stretch pixels into column vector, learning the templates per class.
Linear classifier:
todo: 
1. define a loss function that quantifies our **unhappiness** with the scores across the training data. 
2. come up with a way of efficiently finding the parameters that **minimize** the loss function. 

Loss function:
Multiclass SVM loss, hinge loss, or square loss, it depends on what type of errors you care about

def L_i_vectorized (x, y, W):
    scores = W.dot(x)
    margins = np.maximum(0, scores - scores[y] + 1)
    margins[y] = 0
    loss_i = np.sum(margins)
    return loss_i 

data loss: model predictions should match training data

regularization: model should be "simple", so it works on test data; peneralize
L1 regularization: encourage sparsity, L2 regularization (weight decay); elsatic net; max norm regularization; 
dropout; fancier; batch normalization; stochastic depth; 

Softmax classifier, multinomial logistic regression, cross-entropy
scores = unnormalized log probabilities of the classes. 

Optimization:

Random search:
`bestloss = float('inf')
for num in xrange(1000): 
W = np.random.randn(10,3073)*0.0001 
loss = L(X_train, Y_train, W)
if loss<bestloss:
bestloss = loss
bestW = W
print ( 'in attempt %d the loss was %f, best %f' % (num, loss, bestloss))`

follow the slope:
calculus, analytic gradient, gradient descent, adam optimizer

Stochastic gradient descent:
Aside: image features, color histogram, histogram of oriented gradients, bag of words,
